dev/Kozani/Common/KozaniDvcProtocol.cpp

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) Microsoft Corporation and Contributors.
+// Licensed under the MIT License.
+
+#include <pch.h>
+#include "KozaniDvcProtocol.h"
+
+namespace Microsoft::Kozani::DvcProtocol
+{
+    bool IsEmptyPayloadProtocolDataUnitType(Dvc::ProtocolDataUnit::DataType type)
+    {
+        switch (type)
+        {
+            case Dvc::ProtocolDataUnit::AppTerminationNotice:
+                return true;
+        }
+
+        return false;
+    }
+
+    std::string CreatePDU(UINT64 activityId, Dvc::ProtocolDataUnit::DataType type, const std::string& payload = std::string())
+    {
+        if (!IsEmptyPayloadProtocolDataUnitType(type))
+        {
+            // Payload data of the PDU should not be empty. It catches a failure condition when empty string is returned 
+            // from a failed SerializeAsString call before calling into this method.
+            THROW_HR_IF(E_UNEXPECTED, payload.empty());
```

Comment:
I don't know what exact causes would make it fail. There is no documentation of failure conditions. E_OUTOFMEMORY can be one likely cause. Since it can fail and return an empty string, we should communicate this error to the caller by returning an error code. The caller can deal with it instead of silently using an empty string as if everything is fine, causing issues further down the code path, which will make debugging harder.

Reply:
`E_UNEXPECTED` isn't entirely true (out-of-memory isn't entirely unexpected and unknown).

`KOZANI_E_PDU_SERIALIZATION` would be more useful (custom error code via FACILITY_ITF :-)

`E_UNEXPECTED` should be reserved for the truly unexpected things (typically, we the developers scr*wed up in ways should and could have known to prevent or handle).

Reply:
Added and used KOZANI_E_PDU_SERIALIZATION.